File: typescript/tests/biosignals.test.ts
// biosignals.test.ts
// Verifies the CircleAI.Wearable.Biosignals port: BiosignalKind ordinals, sample
// factory (guid shape + confidence clamp + float rounding), null + recorded
// sources, the windowed aggregator, and the deterministic affect mapper.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { AffectState } from "../src/memory/index";
import {
  BiosignalKind,
  BiosignalAffectMapper,
  BiosignalAggregator,
  NullBiosignalSource,
  RecordedBiosignalSource,
  createBiosignalSample,
  biosignalSample,
} from "../src/wearable/biosignals/index";

const fr = Math.fround;

async function drain<T>(it: AsyncIterable<T>): Promise<T[]> {
  const out: T[] = [];
  for await (const x of it) out.push(x);
  return out;
}

describe("BiosignalKind", () => {
  it("has stable ordinals matching the C# enum", () => {
    assert.equal(BiosignalKind.HeartRate, 0);
    assert.equal(BiosignalKind.HeartRateVariability, 1);
    assert.equal(BiosignalKind.OxygenSaturation, 2);
    assert.equal(BiosignalKind.SleepStage, 5);
    assert.equal(BiosignalKind.Steps, 6);
    assert.equal(BiosignalKind.Unknown, 8);
  });
});

describe("createBiosignalSample", () => {
  it("mints a 32-hex guid id, clamps confidence to [0,1], and frounds value", () => {
    const s = createBiosignalSample(BiosignalKind.HeartRate, 72.3, "bpm", 1.5, false);
    assert.match(s.id, /^[0-9a-f]{32}$/);
    assert.equal(s.confidence, 1); // clamped from 1.5
    assert.equal(s.value, fr(72.3));
    assert.equal(s.unit, "bpm");
    assert.equal(s.isCumulative, false);
    assert.ok(s.measuredAt instanceof Date);

    const low = createBiosignalSample(BiosignalKind.OxygenSaturation, 98, "%", -0.2);
    assert.equal(low.confidence, 0); // clamped from -0.2
  });

  it("positional factory frounds value and confidence", () => {
    const s = biosignalSample("id0", BiosignalKind.BodyTemperature, 36.6, "celsius", 0.9, false, new Date(0));
    assert.equal(s.value, fr(36.6));
    assert.equal(s.confidence, fr(0.9));
  });
});

describe("NullBiosignalSource", () => {
  it("supports nothing and streams nothing", async () => {
    const src = new NullBiosignalSource();
    assert.deepEqual(src.supportedKinds, []);
    assert.equal(await src.isSupportedAsync(BiosignalKind.HeartRate), false);
    assert.deepEqual(await drain(src.streamAsync()), []);
  });
});

describe("RecordedBiosignalSource", () => {
  it("reports its distinct kinds and replays its samples in order", async () => {
    const samples = [
      biosignalSample("s1", BiosignalKind.HeartRate, 60, "bpm", 1, false, new Date("2026-01-01T00:00:00Z")),
      biosignalSample("s2", BiosignalKind.HeartRate, 80, "bpm", 1, false, new Date("2026-01-01T00:00:01Z")),
      biosignalSample("s3", BiosignalKind.Steps, 100, "count", 1, true, new Date("2026-01-01T00:00:02Z")),
    ];
    const src = new RecordedBiosignalSource(samples);
    assert.deepEqual([...src.supportedKinds].sort(), [BiosignalKind.HeartRate, BiosignalKind.Steps].sort());
    assert.equal(await src.isSupportedAsync(BiosignalKind.HeartRate), true);
    assert.equal(await src.isSupportedAsync(BiosignalKind.OxygenSaturation), false);
    assert.deepEqual((await drain(src.streamAsync())).map((s) => s.id), ["s1", "s2", "s3"]);
  });
});

describe("BiosignalAggregator", () => {
  it("aggregates per-kind min/max/mean/count over the window", async () => {
    const now = Date.now();
    const samples = [
      biosignalSample("s1", BiosignalKind.HeartRate, 60, "bpm", 1, false, new Date(now)),
      biosignalSample("s2", BiosignalKind.HeartRate, 80, "bpm", 1, false, new Date(now)),
      biosignalSample("s3", BiosignalKind.HeartRate, 100, "bpm", 1, false, new Date(now)),
      biosignalSample("s4", BiosignalKind.OxygenSaturation, 97, "%", 1, false, new Date(now)),
    ];
    const agg = new BiosignalAggregator(new RecordedBiosignalSource(samples));
    const snap = await agg.snapshotAsync(60_000);
    const hr = snap.stats.get(BiosignalKind.HeartRate);
    assert.ok(hr);
    assert.equal(hr?.sampleCount, 3);
    assert.equal(hr?.min, fr(60));
    assert.equal(hr?.max, fr(100));
    assert.equal(hr?.mean, fr(80)); // (60+80+100)/3
    const spo2 = snap.stats.get(BiosignalKind.OxygenSaturation);
    assert.equal(spo2?.sampleCount, 1);
    assert.equal(spo2?.mean, fr(97));
  });

  it("drops samples older than the window", async () => {
    const now = Date.now();
    const old = biosignalSample("old", BiosignalKind.HeartRate, 999, "bpm", 1, false, new Date(now - 10 * 60_000));
    const fresh = biosignalSample("new", BiosignalKind.HeartRate, 70, "bpm", 1, false, new Date(now));
    const agg = new BiosignalAggregator(new RecordedBiosignalSource([old, fresh]));
    const snap = await agg.snapshotAsync(60_000);
    const hr = snap.stats.get(BiosignalKind.HeartRate);
    assert.equal(hr?.sampleCount, 1);
    assert.equal(hr?.mean, fr(70));
  });

  it("rejects a non-positive window", async () => {
    const agg = new BiosignalAggregator(new NullBiosignalSource());
    await assert.rejects(() => agg.snapshotAsync(0), /Window must be positive/);
  });
});

describe("BiosignalAffectMapper", () => {
  it("ignores low-confidence samples", () => {
    const a = new AffectState();
    const before = a.energy;
    BiosignalAffectMapper.apply(
      biosignalSample("s", BiosignalKind.HeartRate, 140, "bpm", 0.4, false, new Date()),
      a,
    );
    assert.equal(a.energy, before);
  });

  it("applies the high heart-rate rule with float parity", () => {
    const a = new AffectState();
    BiosignalAffectMapper.apply(biosignalSample("s", BiosignalKind.HeartRate, 140, "bpm", 1, false, new Date()), a);
    assert.equal(a.energy, 0.6000000238418579);
    assert.equal(a.uncertainty, 0.25);
  });

  it("applies the elevated and low heart-rate rules", () => {
    const a1 = new AffectState();
    BiosignalAffectMapper.apply(biosignalSample("s", BiosignalKind.HeartRate, 110, "bpm", 1, false, new Date()), a1);
    assert.equal(a1.energy, 0.550000011920929);

    const a2 = new AffectState();
    BiosignalAffectMapper.apply(biosignalSample("s", BiosignalKind.HeartRate, 45, "bpm", 1, false, new Date()), a2);
    assert.equal(a2.energy, 0.44999998807907104);
  });

  it("applies HRV rules (low: uncertainty up + rapport clamped at 0; high: engagement up)", () => {
    const low = new AffectState();
    BiosignalAffectMapper.apply(
      biosignalSample("s", BiosignalKind.HeartRateVariability, 15, "ms", 1, false, new Date()),
      low,
    );
    assert.equal(low.uncertainty, 0.25);
    assert.equal(low.rapport, 0); // 0 - 0.02 clamped to 0

    const high = new AffectState();
    BiosignalAffectMapper.apply(
      biosignalSample("s", BiosignalKind.HeartRateVariability, 70, "ms", 1, false, new Date()),
      high,
    );
    assert.equal(high.engagement, 0.5199999809265137);
  });

  it("applies the low SpO2 rule and leaves other kinds untouched", () => {
    const a = new AffectState();
    BiosignalAffectMapper.apply(
      biosignalSample("s", BiosignalKind.OxygenSaturation, 85, "%", 1, false, new Date()),
      a,
    );
    assert.equal(a.uncertainty, 0.30000001192092896);

    const b = new AffectState();
    const engBefore = b.engagement;
    BiosignalAffectMapper.apply(biosignalSample("s", BiosignalKind.SleepStage, 2, "stage", 1, false, new Date()), b);
    assert.equal(b.engagement, engBefore); // sleep stage does not mutate affect
  });
});

File: kotlin/src/test/kotlin/com/bhengubv/circleai/wearable/biosignals/BiosignalsTest.kt
// BiosignalsTest.kt — verifies the CircleAI.Wearable.Biosignals port against the C# reference.

package com.bhengubv.circleai.wearable.biosignals

import com.bhengubv.circleai.memory.AffectState
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Test
import java.time.Duration
import java.time.Instant
import java.util.UUID
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class BiosignalsTest {

    private fun sample(kind: BiosignalKind, value: Float, conf: Float = 1.0f, at: Instant = Instant.now()) =
        BiosignalSample(UUID.randomUUID(), kind, value, "u", conf, false, at)

    @Test
    fun `kind ordinals are stable`() {
        assertEquals(0, BiosignalKind.HeartRate.ordinal)
        assertEquals(8, BiosignalKind.Unknown.ordinal)
        assertEquals(5, BiosignalKind.SleepStage.ordinal)
    }

    @Test
    fun `create clamps confidence`() {
        assertEquals(1.0f, BiosignalSample.create(BiosignalKind.HeartRate, 70f, "bpm", confidence = 5f).confidence)
        assertEquals(0.0f, BiosignalSample.create(BiosignalKind.HeartRate, 70f, "bpm", confidence = -1f).confidence)
    }

    @Test
    fun `null source supports and emits nothing`() = runTest {
        val s = NullBiosignalSource()
        assertTrue(s.supportedKinds.isEmpty())
        assertFalse(s.isSupportedAsync(BiosignalKind.HeartRate))
        assertTrue(s.streamAsync().toList().isEmpty())
    }

    @Test
    fun `recorded source replays and reports supported kinds`() = runTest {
        val samples = listOf(
            sample(BiosignalKind.HeartRate, 60f),
            sample(BiosignalKind.OxygenSaturation, 97f),
            sample(BiosignalKind.HeartRate, 62f),
        )
        val src = RecordedBiosignalSource(samples)
        assertTrue(src.isSupportedAsync(BiosignalKind.HeartRate))
        assertTrue(src.isSupportedAsync(BiosignalKind.OxygenSaturation))
        assertFalse(src.isSupportedAsync(BiosignalKind.Steps))
        assertEquals(setOf(BiosignalKind.HeartRate, BiosignalKind.OxygenSaturation), src.supportedKinds.toSet())
        assertEquals(3, src.streamAsync().toList().size)
    }

    @Test
    fun `aggregator computes windowed stats`() = runTest {
        val now = Instant.now()
        val src = RecordedBiosignalSource(
            listOf(
                sample(BiosignalKind.HeartRate, 60f, at = now),
                sample(BiosignalKind.HeartRate, 80f, at = now),
                sample(BiosignalKind.HeartRate, 100f, at = now.minus(Duration.ofDays(1))), // outside window
            ),
        )
        val snap = BiosignalAggregator(src).snapshotAsync(Duration.ofMinutes(5))
        val hr = snap.stats[BiosignalKind.HeartRate]!!
        assertEquals(2, hr.sampleCount) // old sample dropped
        assertEquals(60f, hr.min)
        assertEquals(80f, hr.max)
        assertEquals(70f, hr.mean)
        assertFailsWith<IllegalArgumentException> { BiosignalAggregator(src).snapshotAsync(Duration.ZERO) }
    }

    @Test
    fun `affect mapper applies heart-rate rules and confidence gate`() {
        val a = AffectState()
        a.energy = 0.5f
        a.uncertainty = 0.2f
        // Low confidence -> no mutation.
        BiosignalAffectMapper.apply(sample(BiosignalKind.HeartRate, 200f, conf = 0.4f), a)
        assertEquals(0.5f, a.energy)

        // High HR (>130) -> energy +0.10, uncertainty +0.05.
        BiosignalAffectMapper.apply(sample(BiosignalKind.HeartRate, 140f), a)
        approx(0.60f, a.energy)
        approx(0.25f, a.uncertainty)
    }

    @Test
    fun `affect mapper applies hrv and spo2 rules`() {
        val a = AffectState()
        a.uncertainty = 0.2f
        a.rapport = 0.5f
        a.engagement = 0.5f

        BiosignalAffectMapper.apply(sample(BiosignalKind.HeartRateVariability, 10f), a) // <20
        approx(0.25f, a.uncertainty)
        approx(0.48f, a.rapport)

        BiosignalAffectMapper.apply(sample(BiosignalKind.HeartRateVariability, 70f), a) // >60
        approx(0.52f, a.engagement)

        val before = a.uncertainty
        BiosignalAffectMapper.apply(sample(BiosignalKind.OxygenSaturation, 85f), a) // <90
        approx((before + 0.10f).coerceIn(0f, 1f), a.uncertainty)

        // Sleep stage / other kinds -> no affect change (but lastUpdated is touched).
        val u = a.uncertainty
        BiosignalAffectMapper.apply(sample(BiosignalKind.SleepStage, 2f), a)
        approx(u, a.uncertainty)
    }
}

private fun approx(expected: Float, actual: Float, tol: Float = 1e-6f) =
    kotlin.test.assertTrue(kotlin.math.abs(expected - actual) <= tol, "expected $expected got $actual")
